pkg/controller/rollout/rollout_canary.go

```diff
@@ -0,0 +1,504 @@
+			canaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			canaryStatus.CurrentStepState = v1alpha1.CanaryStepStateCompleted
+		}
+		klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+			canaryStatus.CurrentStepIndex, v1alpha1.CanaryStepStateReady, canaryStatus.CurrentStepState)
+		// canary completed
+	case v1alpha1.CanaryStepStateCompleted:
+		klog.Infof("rollout(%s/%s) run canary strategy, and state(%s)", c.Rollout.Namespace, c.Rollout.Name, v1alpha1.CanaryStepStateCompleted)
+	}
+
+	return nil
+}
+
+func (m *canaryReleaseManager) doCanaryUpgrade(c *util.RolloutContext) (bool, error) {
+	// verify whether batchRelease configuration is the latest
+	steps := len(c.Rollout.Spec.Strategy.Canary.Steps)
+	canaryStatus := c.NewStatus.CanaryStatus
+	cond := util.GetRolloutCondition(*c.NewStatus, v1alpha1.RolloutConditionProgressing)
+	cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and upgrade workload to new version", canaryStatus.CurrentStepIndex, steps)
+	c.NewStatus.Message = cond.Message
+	// run batch release to upgrade the workloads
+	done, br, err := m.runBatchRelease(c.Rollout, getRolloutID(c.Workload), canaryStatus.CurrentStepIndex, c.Workload.IsInRollback)
+	if err != nil {
+		return false, err
+	} else if !done {
+		return false, nil
+	}
+	if br.Status.ObservedReleasePlanHash != util.HashReleasePlanBatches(&br.Spec.ReleasePlan) ||
+		br.Generation != br.Status.ObservedGeneration {
+		klog.Infof("rollout(%s/%s) batchRelease status is inconsistent, and wait a moment", c.Rollout.Namespace, c.Rollout.Name)
+		return false, nil
+	}
+	// check whether batchRelease is ready(whether new pods is ready.)
+	if br.Status.CanaryStatus.CurrentBatchState != v1alpha1.ReadyBatchState ||
+		br.Status.CanaryStatus.CurrentBatch+1 < canaryStatus.CurrentStepIndex {
+		klog.Infof("rollout(%s/%s) batchRelease status(%s) is not ready, and wait a moment", c.Rollout.Namespace, c.Rollout.Name, util.DumpJSON(br.Status))
+		return false, nil
+	}
+	m.recorder.Eventf(c.Rollout, corev1.EventTypeNormal, "Progressing", fmt.Sprintf("upgrade step(%d) canary pods with new versions done", canaryStatus.CurrentStepIndex))
+	klog.Infof("rollout(%s/%s) batch(%s) state(%s), and success",
+		c.Rollout.Namespace, c.Rollout.Name, util.DumpJSON(br.Status), br.Status.CanaryStatus.CurrentBatchState)
+	return true, nil
+}
+
+func (m *canaryReleaseManager) doCanaryMetricsAnalysis(c *util.RolloutContext) (bool, error) {
+	// todo
+	return true, nil
+}
+
+func (m *canaryReleaseManager) doCanaryPaused(c *util.RolloutContext) (bool, error) {
+	canaryStatus := c.NewStatus.CanaryStatus
+	currentStep := c.Rollout.Spec.Strategy.Canary.Steps[canaryStatus.CurrentStepIndex-1]
+	steps := len(c.Rollout.Spec.Strategy.Canary.Steps)
+	cond := util.GetRolloutCondition(*c.NewStatus, v1alpha1.RolloutConditionProgressing)
+	// need manual confirmation
+	if currentStep.Pause.Duration == nil {
```

Comment:
If `currentSetp == "100%"`, should we still need this manual confirmation?

Reply:
The situation is a bit complicated, because there are two fields, weight and replicas, if one of them is 100% and the other is not, it is difficult to judge.